{% extends "layout.html" %}
{% block head %}
    <!-- Render context vars in JS here -->
    <script type="text/javascript">
        window.icd = {{ icd|tojson|safe }};
        window.page_name = window.icd[0].icd_id;
    </script>
    <script type="text/javascript" src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <script>
        filtered_variants = _.filter(window.table_variants, function(d) {
            return d.filter == 'PASS';
        });
        $('#number_of_variants').replaceWith(filtered_variants.length);
        lof_filtered_variants = _.filter(window.table_variants, function(d) {
            return d.filter == 'PASS' && d.category == 'lof_variant';
        });$('#number_of_lof_variants').replaceWith(lof_filtered_variants.length);
    </script>

    <style>
        table {
            border-collapse: collapse;
            border: 1px black solid;
            font: sans-serif;
/*
            border: 2px black solid;
            font: 12px sans-serif;
*/
        }
        td {
            border: 1px black solid;
            padding: 5px;
        }

        h3{background-color:#826aa5;color:#fff;padding:3px;}
        h3 span{float:right;font-size:28px;font-weight:normal;padding:1px;}
        h3 span a{color:#fff;text-decoration:none;}

        @font-face {
        font-family: 'Glyphicons Halflings';
        src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot');
        src: url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.eot?#iefix') format('embedded-opentype'),
            url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff2') format('woff2'),
            url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.woff') format('woff'),
            url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.ttf') format('truetype'),
            url('//cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular') format('svg');
        }

        .glyphicon {
            position: relative;
            top: 1px;
            display: inline-block;
            font: normal normal 16px/1 'Glyphicons Halflings';
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            margin-right: 4px;
        }
        /* Add icons you will be using below */
            .glyphicon-fire:before {
            content: '\e104';
        }
        .glyphicon-eye-open:before {
            content: '\e105';
        }
        .glyphicon-cloud-download:before {
            content: '\e197';
        }
    </style>
{% endblock %}

{% block body %}
<div class="container">
    <div class="row">
        <div class="col col-md-12">
            <h1>Sparse polygenic risk scores (PRS) with BASIL / snpnet</h1>
            <hr/>
            <h5>Version 1.1, 01.16.2021</h5>
            <p>Yosuke Tanigawa and Manuel A. Rivas</p>
	        <p>Note: this page is under active development.</p>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h1>Phenotype: {{icd[0].Name}}</h1>
            <p><a href="/prs">Go back to the PRS top page</a></p>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-6">
            <dl class="dl-horizontal" style="margin-bottom: 0px;">
                <dt>Phenotype code:</dt>
                <dt>Case Count or N:</dt>
                <dt>Single variant:</dt>
                <dd>{{ icd[0].icd }}</dd>
                <dd>{{ icd[0].Case }}</dd>
                <dd><a href="/{{namespace}}/coding/{{ icd[0].icd}}">{{ icd[0].icd.replace('ICD','') }}</a></dd>
            </dl>
        </div>
        <div class="col col-md-6">
            <a href="/static/PRS_map/{{ icd[0].icd}}.tsv" class="btn btn-block btn-primary">
                <span class="glyphicon glyphicon-cloud-download"></span> {{icd[0].Name}} PRS weights
            </a>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h2 align="center">Summary of the predictive performance</h2>
            <img class="img-responsive" src={{ snpnet_plot |tojson|safe }} alt={{ icd_str|tojson|safe }}>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h2 align="center">Coefficients of the PRS model</h2>
            <br>
            <div id="containergenome2"></div>
            <br>
            Note: only <p id="someText" style="display:inline"></p> variants in active path for Lasso are included in the coefficient plot.
            <br>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h2 align="center">Predictive performance of the PRS model</h2>
            <br>
            <div id="containerprseval" style="font-size:120%;"></div>
            <br>
            <hr/>

            <h3>Description of the columns</h3>
            <ul>
                <li>Trait: a trait in UK Biobank.</li>
                <li>Genotype-only: the predictive performence (r2 or AUC) of the predictive model with genotype features alone.</li>
                <li>Covariate-only: the predictive performence (r2 or AUC) of the predictive model with covariates (age, sex, and genotype PCs) alone.</li>
                <li>Full model: the predictive performence (r2 or AUC) of the full model that have both genotypes and covariates alone.</li>
                <li>diff (Full - Covars): the incremental predictive performence (r2 or AUC) of the full model against the covariate-only model, i.e. the difference between the values in "Full" and "Covar" columns.</li>
                <li># variants: the number of genetic variants included in the model</li>
            </ul>
            <br>
            <hr/>


        </div>
    </div>

</div>
{% endblock %}

{% block endofbody %}
<script type="text/javascript" charset="utf-8">
    d3.text( "{{ snpnet_eval }}", function(data) {
        var parsedCSV = d3.tsv.parseRows(data);
	// manually change the header line
	parsedCSV[0][0] = 'GBE_ID' ;
	parsedCSV[0][1] = 'Split / ancestry' ;
	parsedCSV[0][2] = 'Genotype-only' ;
	parsedCSV[0][3] = 'Covariate-only' ;
	parsedCSV[0][4] = 'Full model' ;
	parsedCSV[0][5] = 'diff (Full - Covars)' ;
	for (i = 1; i < parsedCSV.length; i++) {
	// round the number to 2 dicimal places
	    parsedCSV[i][2] = Number(parsedCSV[i][2]).toFixed(2) ;
	    parsedCSV[i][3] = Number(parsedCSV[i][3]).toFixed(2) ;
	    parsedCSV[i][4] = Number(parsedCSV[i][4]).toFixed(2) ;
	    parsedCSV[i][5] = Number(parsedCSV[i][5]).toFixed(2) ;
	}
        var container = d3.select("#containerprseval")
            .append("table")
	    .attr("class", "table")

            .selectAll("tr")
                .data(parsedCSV).enter()
                .append("tr")

            .selectAll("td")
                .data(function(d) { return d; }).enter()
                .append("td")
                .text(function(d) { return d; });
    });
</script>

<script>
    var cutoff = 7;
    var activeset = 0;
    var activenum = 1;
    // manhattan plot code obtained/edited/adapted from : https://github.com/chengsoonong/rede/tree/master/Rede/manhattanplot
    d3.select("#containergenome2")
            .style("height", "100%")
            .style("overflow", "auto");

    var chrom_length = 0;
    var chrom_acum_length = new Array();
    var chromLength = new Array(
        249250621, 243199373, 198022430, 191154276,
        180915260, 171115067, 159138663, 146364022,
        141213431, 135534747, 135006516, 133851895,
        115169878, 107349540, 102531392, 90354753,
        81195210, 78077248, 59128983, 63025520,
        48129895, 51304566, 155270560, 59373566);

    //this initializes chrom_length and chrom_acum_length to be used in manhattan plot
    for (var i = 0; i < chromLength.length; i++) {
        chrom_length = chrom_length + chromLength[i];
        chrom_acum_length.push(chrom_length);
    }

    posarr = new Array();
    data_low = new Array();
    data_high = new Array();
    pvalarr = new Array();
    idarr = new Array();
    data_weight_pvalue = [];
    d3.tsv("/static/PRS_map/" + icd[0].icd + ".tsv",function(error,data){
        data.forEach(function(d) {
    // Fill two seperate arrays, one with points greater than the -log10(pvalue) threshold, and another
    // with points lower than the threshold defined in cutoff
            d.chrom = +d["CHROM"];
            d.pos = +d["POS"];
            d.beta = +d["BETA"];
            if (d.beta === 0) {
                return;
            }
            activeset = activeset + 1;
            d.id = d["CHROM"] + ":" + d["POS"] + "-" +  d["REF"] + "-" + d["ALT"];
            if (d.chrom == 1) {
                if (d.beta > -.01 && d.beta < .01) {
                    data_low.push( [d.pos, d.beta, d.id, d.chrom]);
                } else {
                    data_high.push([d.pos, d.beta, d.id, d.chrom]);
                }
            } else {
                if (d.beta > -.01 && d.beta < .01) {
                    data_low.push( [d.pos + chrom_acum_length[d.chrom - 2], d.beta, d.id, d.chrom]);
                } else {
                    data_high.push([d.pos + chrom_acum_length[d.chrom - 2],	d.beta, d.id, d.chrom]);
                }
            }
            data_weight_pvalue.push(d.beta);
        })

        // Merge the both high and low significance points together into a single array
        all_data = data_low.concat(data_high);
        var min_pvalue = d3.max([-3, d3.min([-.005, d3.min(data_weight_pvalue)])]);
        var max_pvalue = d3.max([.005, d3.max(data_weight_pvalue)]);
        // var which defined the extra space in the bottom and the top of the -log(p-value) axis dynamically to the dataset
        var extend_scale = (max_pvalue - min_pvalue) *0.05;
        ix_1 = 0;
        ix_2 = chrom_length;
        iy_1 = min_pvalue - extend_scale;
        iy_2 = max_pvalue + extend_scale;

        //create the manhattan plot
        x1 = ix_1;
        x2 = ix_2;
        y1 = iy_1;
        y2 = iy_2;

        var margin = {
            top: 30,
            right: 50,
            bottom: 20,
            left: 60
        };

        var colorScale = d3.scale.log()
            .domain([d3.min(all_data, function(d) {
                return parseInt(d[3]);
            }), d3.max(all_data, function(d) {
                return parseInt(d[3]);
            })])
            .interpolate(d3.interpolateHsl)
            .range(["#B3995D", "#8C1515"]);

        var w = 1100 - margin.left - margin.right; //900;
        var h = 500 - margin.top - margin.bottom; //600;

        //Create scale functions
        var xScale = d3.scale.linear()
            .domain([x1, x2])
            .range([0, w]);

        var yScale = d3.scale.linear()
            .domain([y1, y2])
            .range([h, 0]);

        var array_test1 = [""];
        var array_test2 = [0];

        for (var i = 0; i < chrom_acum_length.length; i++) {
            var num = i + 1;
            array_test1.push("chr" + num);
            array_test2.push(xScale(chrom_acum_length[i]));
        }
        //  console.log(x1,x2,y1,y2);

        //Create scale top
        var xScale_top = d3.scale.ordinal()
            .domain(array_test1)
            .range(array_test2);

        //Define X axis top
        var xAxis_top = d3.svg.axis()
            .scale(xScale_top)
            .orient("top")
            .ticks(0);

        //Define X axis
        var xAxis = d3.svg.axis()
            .scale(xScale)
            .orient("bottom");

        //Define Y axis
        var yAxis = d3.svg.axis()
            .scale(yScale)
            .orient("left")
            .ticks(5);

        // Define tooltip
        var tooltip = d3.select("#containergenome2")
            .append("div")
            .style("position", "fixed")
            .style("z-index", "0")
            .style("visibility", "hidden")
            .text("a simple tooltip");

        //Create SVG element
        var svg = d3.select("#containergenome2")
            .append("svg")
            .attr("width", w + margin.left + margin.right)
            .attr("height", h + margin.top + margin.bottom)
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

        // Vertical chromosome lines
        var line_chrom = svg.selectAll("line")
            .data(chrom_acum_length)
            .enter()
            .append("line")
            .attr("class", "linechrom")
            .attr("x1", function(d) {
                return xScale(d);
            })
            .attr("y1", 0)
            .attr("x2", function(d) {
                return xScale(d);
            })
            .attr("y2", h)
            .attr("stroke-width", 1)
            .attr("stroke-dasharray", 5)
            .style("stroke", "black") //stroke-dasharray="5"
            .style("opacity", 0.5);

        //Create circles for the low significance points
        var circle_low = svg.selectAll("circle_low")
            .data(data_low)
            .enter()
            .append("circle")
            .attr("cx", function(d) {
                return xScale(parseFloat(d[0]));
            })
            .attr("cy", function(d) {
                return yScale(parseFloat(d[1]));
            })
            .attr("r", 1)
            .style("fill", function(d) {
                return "#D3D3D3";
            });

        // Create circles for the high significance points
        var circle_high = svg.selectAll("circle_high")
            .data(data_high)
            .enter()
            .append("circle")
            .attr("cx", function(d) {
                return xScale(parseFloat(d[0]));
            })
            .attr("cy", function(d) {
                return yScale(parseFloat(d[1]));
            })
            .attr("r", function(d){
                return d3.max([2,1.5*parseFloat(d[1])]);
                })
            .on("mouseover", function(d){return tooltip.style("top", (event.pageY ) + "px").style("left", (event.pageX) + "px").style("visibility", "visible").text("variant id: " + d[2] + "; " + "Lasso coefficient: " + Number((d[1]).toFixed(4)));})
            .on("mousemove", function(){return tooltip.style("top", (event.pageY)+"px").style("left",(event.pageX)+"px");})
            .on("mouseout", function(){return tooltip.style("visibility", "hidden");})
            .on("click", function(d){
            window.open("https://biobankengine.stanford.edu/{{ namespace }}/awesome?query=" + d[2], '_blank');
            })
            .style("fill", function(d) {
                return colorScale(parseInt(d[3]));
            });

        //Create X axis
        svg.append("g")
            .attr("class", "manaxis").attr("font-size", "10px")
            .attr("transform", "translate(0," + (h) + ")")
            .call(xAxis);

        svg.append("g").attr("transform", "translate(0," + 0 + ")")
            .attr("class", "xt axis")
            .call(xAxis_top);

        svg.selectAll(".xt text") // select all the text elements for the xaxis
            .attr("transform", function(d) {
                return "translate(" + this.getBBox().height + "," + this.getBBox().height * -0.5 + ")rotate(-45)";
            });

        //Create Y axis
        svg.append("g")
            .attr("class", "manaxis")
            .call(yAxis)
            .append("text")
            .attr("class", "manlabel")
            .attr("transform", "rotate(-90)")
            .attr("x", -10)
            .attr("y", -50)
            .attr("dy", ".71em")
            .style("text-anchor", "end").attr("font-size", "17px")
            .text("Lasso coefficient");
        document.getElementById("someText").innerHTML = activeset.toString();
    });
</script>
<!--
	 <div id="containergenome3">

        <script>
        var cutoff = 7;
	var activeset2 = 0;
        var activenum2 = 1;


    // manhattan plot code obtained/edited/adapted from : https://github.com/chengsoonong/rede/tree/master/Rede/manhattanplot
    d3.select("#containergenome3")
                .style("height", "100%")
                .style("overflow", "auto");

    var chrom_length = 0;
	var chrom_acum_length = new Array();
	var chromLength = new Array(249250621, 243199373, 198022430, 191154276,
    180915260, 171115067, 159138663, 146364022,
    141213431, 135534747, 135006516, 133851895,
    115169878, 107349540, 102531392, 90354753,
    81195210, 78077248, 59128983, 63025520,
    48129895, 51304566, 155270560, 59373566);

	//this initializes chrom_length and chrom_acum_length to be used in manhattan plot
	for (var i = 0; i < chromLength.length; i++) {
  	  chrom_length = chrom_length + chromLength[i];
   	 chrom_acum_length.push(chrom_length);
	}


	posarr = new Array();
	data_low = new Array();
        data_high = new Array();
 	pvalarr = new Array();
        idarr = new Array();
	data_weight_pvalue = [];
        d3.tsv("/static/prs/" + icd[0].icd + ".coxnet.tsv",function(error,data){
    data.forEach(function(d) {
	// Fill two seperate arrays, one with points greater than the -log10(pvalue) threshold, and another
	// with points lower than the threshold defined in cutoff
               d.chrom = +d["CHROM"];
	       d.pos = +d["POS"];
	       d.beta = +d["BETA"];
if (d.beta === 0) {
    return;
  }
               activeset2 = activeset2 + 1;
	       d.id = d["ID"];
         	if (d.chrom == 1) {
                if (d.beta < -1000) {
                        data_low.push([d.pos, d.beta, d.id,
                    	     d.chrom]);
                } else {
					data_high.push([d.pos, d.beta, d.id,
			      d.chrom]);
				}
			} else {
				if (d.beta < -1000) {
					data_low.push([d.pos + chrom_acum_length[d.chrom - 2],
						d.beta, d.id,
						d.chrom]);
				} else {
						data_high.push([d.pos + chrom_acum_length[d.chrom - 2],
							d.beta, d.id,
							d.chrom]);


	}

        }
        		        	data_weight_pvalue.push(d.beta);


		})

	// Merge the both high and low significance points together into a single array
    all_data = data_low.concat(data_high);
    var min_pvalue = d3.max([-3, d3.min([-.005, d3.min(data_weight_pvalue)])]);
    var max_pvalue = d3.max([.005, d3.max(data_weight_pvalue)]);
    // var which defined the extra space in the bottom and the top of the -log(p-value) axis dynamically to the dataset
    var extend_scale = (max_pvalue - min_pvalue) *0.05;
    ix_1 = 0;
    ix_2 = chrom_length;
    iy_1 = min_pvalue - extend_scale;
    iy_2 = max_pvalue + extend_scale;

	//create the manhattan plot
	x1 = ix_1;
	x2 = ix_2;
	y1 = iy_1;
	y2 = iy_2;

    var margin = {
        top: 30,
        right: 50,
        bottom: 20,
        left: 60
    };
    var colorScale = d3.scale.log()
        .domain([d3.min(all_data, function(d) {
            return parseInt(d[3]);
        }), d3.max(all_data, function(d) {
            return parseInt(d[3]);
        })])
        .interpolate(d3.interpolateHsl)
        .range(["#DAF7A6", "#900C3F"]);

    var w = 1100 - margin.left - margin.right; //900;
    var h = 500 - margin.top - margin.bottom; //600;


    //Create scale functions
    var xScale = d3.scale.linear()
        .domain([x1, x2])
        .range([0, w]);

    var yScale = d3.scale.linear()
        .domain([y1, y2])
        .range([h, 0]);

    var array_test1 = [""];
    var array_test2 = [0];

    for (var i = 0; i < chrom_acum_length.length; i++) {
        var num = i + 1;
        array_test1.push("chr" + num);
        array_test2.push(xScale(chrom_acum_length[i]));
    }
  //  console.log(x1,x2,y1,y2);

    //Create scale top
    var xScale_top = d3.scale.ordinal()
        .domain(array_test1)
        .range(array_test2);

    //Define X axis top
    var xAxis_top = d3.svg.axis()
        .scale(xScale_top)
        .orient("top")
        .ticks(0);

    //Define X axis
    var xAxis = d3.svg.axis()
        .scale(xScale)
        .orient("bottom");

    //Define Y axis
    var yAxis = d3.svg.axis()
        .scale(yScale)
        .orient("left")
        .ticks(5);

    // Define tooltip
    var tooltip = d3.select("#containergenome3")
	    .append("div")
	    .style("position", "fixed")
	    .style("z-index", "0")
	    .style("visibility", "hidden")
	    .text("a simple tooltip");

    //Create SVG element
    var svg = d3.select("#containergenome3")
        .append("svg")
        .attr("width", w + margin.left + margin.right)
        .attr("height", h + margin.top + margin.bottom)
        .append("g")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

	// Vertical chromosome lines
    var line_chrom = svg.selectAll("line")
        .data(chrom_acum_length)
        .enter()
        .append("line")
        .attr("class", "linechrom")
        .attr("x1", function(d) {
            return xScale(d);
        })
        .attr("y1", 0)
        .attr("x2", function(d) {
            return xScale(d);
        })
        .attr("y2", h)
        .attr("stroke-width", 1)
        .attr("stroke-dasharray", 5)
        .style("stroke", "black") //stroke-dasharray="5"
        .style("opacity", 0.5);

    //Create circles for the low significance points
    var circle_low = svg.selectAll("circle_low")
        .data(data_low)
        .enter()
        .append("circle")
        .attr("cx", function(d) {
            return xScale(parseFloat(d[0]));
        })
        .attr("cy", function(d) {
            return yScale(parseFloat(d[1]));
        })
        .attr("r", 2)
        .style("fill", function(d) {
            return "#D3D3D3";
        });

	// Create circles for the high significance points
    var circle_high = svg.selectAll("circle_high")
        .data(data_high)
        .enter()
        .append("circle")
        .attr("cx", function(d) {
            return xScale(parseFloat(d[0]));
        })
        .attr("cy", function(d) {
            return yScale(parseFloat(d[1]));
        })
        .attr("r", function(d){
             return d3.max([2,1.5*parseFloat(d[1])]);
             })
        .on("mouseover", function(d){return tooltip.style("top", (event.pageY - 0.001) + "px").style("left", (event.pageX+.001) + "px").style("visibility", "visible").text("variant id: " + d[2] + "; " + "Lasso coefficient: " + Number((d[1]).toFixed(4)));})
        .on("mousemove", function(){return tooltip.style("top", (event.pageY-.001)+"px").style("left",(event.pageX+.001)+"px");})
        .on("mouseout", function(){return tooltip.style("visibility", "hidden");})
        .on("click", function(d){
       	window.open("https://biobankengine.stanford.edu/awesome?query=" + d[2], '_blank');
    	})
        .style("fill", function(d) {
            return colorScale(parseInt(d[3]));
        });

    //Create X axis
    svg.append("g")
        .attr("class", "manaxis").attr("font-size", "10px")
        .attr("transform", "translate(0," + (h) + ")")
        .call(xAxis);

    svg.append("g").attr("transform", "translate(0," + 0 + ")")
        .attr("class", "xt axis")
        .call(xAxis_top);

    svg.selectAll(".xt text") // select all the text elements for the xaxis
        .attr("transform", function(d) {
            return "translate(" + this.getBBox().height + "," + this.getBBox().height * -0.5 + ")rotate(-45)";
        });

    //Create Y axis
    svg.append("g")
        .attr("class", "manaxis")
        .call(yAxis)
        .append("text")
        .attr("class", "manlabel")
        .attr("transform", "rotate(-90)")
        .attr("x", -10)
        .attr("y", -50)
        .attr("dy", ".71em")
        .style("text-anchor", "end").attr("font-size", "17px")
        .text("Lasso coefficient");
    document.getElementById("someText2").innerHTML = activeset2.toString();
});
</script>
</div>
<br>
Note: only <p id="someText2" style="display:inline"></p> variants in active path for Lasso are included in th coefficient plot.
<br>

<script>
percarr = new Array();
dataperc_low = new Array();
dataperc_high = new Array();
prevarr = new Array();
ncasearr = new Array();
nconarr = new Array();

d3.csv("/static/prs/" + icd[0].icd + ".csv",function(error,data){
    data.forEach(function(d) {
            console.log(d);

 if( d.id == "OR"){
               percarr.push(+d["5"], +d["15"], +d["25"], +d["35"], +d["45"], +d["55"], +d["65"], +d["75"], +d["85"], +d["95"]);
            }
            if( d.id == "OR_l"){
               dataperc_low.push(+d["5"], +d["15"], +d["25"], +d["35"], +d["45"], +d["55"], +d["65"], +d["75"], +d["85"], +d["95"]);
            }
            if( d.id == "OR_u"){
               dataperc_high.push(+d["5"], +d["15"], +d["25"], +d["35"], +d["45"], +d["55"], +d["65"], +d["75"], +d["85"], +d["95"]);
            }
            if( d.id == "Prevalence"){
               prevarr.push(+d["5"], +d["15"], +d["25"], +d["35"], +d["45"], +d["55"], +d["65"], +d["75"], +d["85"], +d["95"]);
             }
            if( d.id == "n_cases"){
               ncasearr.push(+d["5"], +d["15"], +d["25"], +d["35"], +d["45"], +d["55"], +d["65"], +d["75"], +d["85"], +d["95"]);
             }
            if( d.id == "n_conts"){
               nconarr.push(+d["5"], +d["15"], +d["25"], +d["35"], +d["45"], +d["55"], +d["65"], +d["75"], +d["85"], +d["95"]);
             }

});

var perchigh = new Array();
var perclow = new Array();
perchigh = dataperc_high.map(function(x, index){
     return x - percarr[index];
  });

perclow = percarr.map(function(x, index){
     return x - dataperc_low[index];
  });

var trace1 =  {
    x: ['5', '15', '25', '35', '45', '55', '65', '75', '85', '95'],
//    x: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    y: percarr,
    error_y: {
      type: 'data',
      symmetric: false,
      array: perchigh,
      arrayminus: perclow
    },
    name: 'odds ratio',
    type: 'scatter',
   yaxis: 'y2'
  };

var trace2 = {
    x: ['5', '15', '25', '35', '45', '55', '65', '75', '85', '95'],
  y: ncasearr,
  name: 'Cases',
  type: 'bar',
  yaxis: 'y',
  opacity: 0.6
};


var trace3 = {
    x: ['5', '15', '25', '35', '45', '55', '65', '75', '85', '95'],
  y: nconarr,
  name: 'Controls',
  type: 'bar',
  yaxis: 'y',
  opacity: 0.6
};

var data = [
trace1, trace2, trace3
];

var layout = {
  barmode: 'group',
  title: {
    text:'Odds ratio plot - Polygenic Risk Score percentiles',
    xref: 'paper',
    x: 0.05,
  },
  xaxis: {
    title: {
      text: 'Percentile',
    },
  },
  yaxis2: {
    title: 'Odds ratio',
    overlaying: 'y',
    side: 'left'
  },
  yaxis: {
   title: 'Number of individuals',
      side: 'right'
}
};


console.log(percarr.length);
console.log(dataperc_high.length);
console.log(data);
console.log(perclow);
console.log(perchigh);
console.log(dataperc_high);
console.log(dataperc_low);


var dataprev = [
  {
    x: ['5', '15', '25', '35', '45', '55', '65', '75', '85', '95'],
//    x: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    y: prevarr,
    mode: 'markers',
    type: 'scatter'
  }
];

var prevlayout = {
  title: {
    text:'Prevalence plot - Polygenic Risk Score percentiles',
    xref: 'paper',
    x: 0.05,
  },
  xaxis: {
    title: {
      text: 'Percentile',
    },
  },
  yaxis: {
    title: {
      text: 'Prevalence',
    }
  }
};

Plotly.newPlot('orplot', data, layout, {showSendToCloud: true});
Plotly.newPlot('prevplot', dataprev, prevlayout, {showSendToCloud: true});
});
</script>
<h3>Odds ratio plot</h3>
<div id="orplot">
</div>



<h3>Prevalence plot</h3>
<div id="prevplot">
</div>
<h3>Survival curves</h3>
<div>
    <a href="https://plot.ly/~marivascruz/690/?share_key=k7l7ladOrcxhljrZ6qWnmd" target="_blank" title="r-gpsurvival-miphs" style="display: block; text-align: center;"><img src="https://plot.ly/~marivascruz/690.png?share_key=k7l7ladOrcxhljrZ6qWnmd" alt="r-gpsurvival-miphs" style="max-width: 100%;width: 600px;"  width="600" onerror="this.onerror=null;this.src='https://plot.ly/404.png';" /></a>
    <script data-plotly="marivascruz:690" sharekey-plotly="k7l7ladOrcxhljrZ6qWnmd" src="https://plot.ly/embed.js" async></script>
</div>

<h3>Forest plot</h3>
<div>
<iframe width="900" height="800" frameborder="0" scrolling="no" src="//plot.ly/~marivascruz/696.embed"></iframe>
</div>

</div>
-->
{% endblock %}
