{% extends "layout.html" %}
{% block head %}
<style>
    .img-fluid {
        max-width: 100%;
        height: auto
      }
</style>
{% endblock %}
{% block body %}
<div class="container">
    <div class="row">
        <div class="col col-md-12">
            <h1>chr3 region analysis - {{ HGI_cc }}</h1>
            <hr/>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <ul>
                <li>Go back to the <a href="/covid19HGI">top page</a></li>
                {% for HGI_suffix in _HGI_suffices %}
                {% if HGI_suffix == HGI_sx %}
                <li>PRS PheWAS results for <a href="/covid19HGI/PRS_PheWAS/{{ HGI_sx }}">{{ HGI_suffix }} dataset</a></li>.
                <ul>
                    {% for HGI_case_control in _HGI_case_controls %}
                    {% if HGI_case_control == HGI_cc %}
                    <li>The additional analyses on chr3 for {{ HGI_case_control }} (this page)</li>
                    {% else %}
                    <li>The additional analyses on chr3 for <a href="/covid19HGI/PRS_PheWAS/{{ HGI_sx }}/{{ HGI_case_control }}">{{ HGI_case_control }}</a></li>
                    {% endif %}
                    {% endfor %}
                </ul>
                {% else %}
                <li>PRS PheWAS results for <a href="/covid19HGI/PRS_PheWAS/{{ HGI_suffix }}">{{ HGI_suffix }} dataset</a></li>
                {% endif %}
                {% endfor %}
            </ul>
            <hr/>
        </div>
    </div>

    {% for clumpp in _clump_p1s %}
    <div class="row">
        <div class="col col-md-12">
            <h2>LD clumping p-value threshold: {{ clumpp }}</h2>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-6">
            <img class="img-fluid" src="{{ url_for('static', filename='covid19HGI/fig/UKB_PRS_PheWAS' )}}/{{ HGI_sx }}/{{ _HGI_v }}_{{ HGI_cc }}_{{ HGI_sx }}.clump{{ clumpp }}.png" />
            <p>
                The PRS-PheWAS analysis for COVID-19 HGI case/control definition, {{ HGI_cc }}. We constructed LD clumping based PRS (clumping p-value threshold: {{ clumpp }}) and performed PRS-PheWAS analysis using phenotypes in UK Biobank (y-axis). We display the significance (x-axis) and the direction (shape) of the association.
            </p>
        </div>
    </div>

    <div class="row">
        <div class="col col-md-12">
            <h3>The effects of chr3 region (with chemokine genes)</h3>
            <img class="img-fluid" src="{{ url_for('static', filename='covid19HGI/fig/remove_chr3_chemokine' )}}/{{ HGI_sx }}/{{ _HGI_v }}_{{ HGI_cc }}_{{ HGI_sx }}.clump{{ clumpp }}.remove_chr3_chemokine.png" />
            <p>
                We repeated the PRS-PheWAS analysis without the COVID-19 associations in the chr3 regions. We compare the p-value (left) and the BETA[SE] (right) of the PRS-PheWAS associations with (x-axis) and without (y-axis) the chr3 region (<a href="/covid19HGI#methods">Methods</a>).
            </p>
        </div>
    </div>

    {% if HGI_sx == 'eur_leave_ukbb_23andme' %}
    <div class="row">
        <div class="col col-md-12">
            <h3>Pairwise comparison of GWAS association statistics</h3>
            <p>
                Focusing on the clumped variants, we compare the GWAS association summary statistics between that of COVID-19 and that of a trait in UK Biobank (<a href="/covid19HGI#methods">Methods</a>).
            </p>
        </div>
    </div>

    {% for GBE_ID in _GBE_IDs %}
    <div class="row">
        <div class="col col-md-12">
            <h4>{{ GBE_ID }}</h4>
            <img class="img-fluid" src="{{ url_for('static', filename='covid19HGI/fig/UKB_PRS_PheWAS_follow_up_GWAS' )}}/{{ HGI_sx }}/{{ HGI_cc }}/{{ _HGI_v }}_{{ HGI_cc }}_{{ HGI_sx }}.clump{{ clumpp }}.{{ GBE_ID }}.png" />
            <p>
                We applied LD clumping (p-value threshold: {{ clumpp }}) to the meta-analyzed summary statistics from COVID-19 HGI ({{ HGI_cc }}).
                Focusing on the identified lead variants, we compared the GWAS association summary statistics, p-value (left) and BETA[SE] (right), for the COVID-19 HGI trait (x-axis) and the UKB trait ({{ GBE_ID }}) (y-axis).
                The genetic variants are colored orange if the variant is in the 3 Mbp window in the chr3 region and blue otherwise.
            </p>
        </div>
    </div>
    {% endfor %}
    {% endif %}

    <div class="row">
        <div class="col col-md-12">
            <hr/>
        </div>
    </div>
    {% endfor %}

</div>
{% endblock %}

